File: src/common/python/drawing.py
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
Drawing Utilities for Visualization
"""

import cv2
import numpy as np
from typing import List, Tuple, Optional, Dict, Any


def generate_colors(num_colors: int, seed: int = 42) -> List[Tuple[int, int, int]]:
    """
    Generate a list of distinct colors for visualization.

    Args:
        num_colors: Number of colors to generate
        seed: Random seed for reproducibility

    Returns:
        List of RGB color tuples
    """
    np.random.seed(seed)
    colors = []
    for _ in range(max(num_colors, 100)):
        color = tuple(map(int, np.random.randint(0, 255, 3)))
        colors.append(color)
    return colors


def draw_bbox(image: np.ndarray,
              bbox: List[float],
              label: str = "",
              color: Tuple[int, int, int] = (0, 255, 0),
              thickness: int = 2) -> np.ndarray:
    """
    Draw a single bounding box on an image.

    Args:
        image: Input image
        bbox: Bounding box [x1, y1, x2, y2]
        label: Label text to display
        color: Box color (BGR)
        thickness: Line thickness

    Returns:
        Image with drawn bounding box
    """
    x1, y1, x2, y2 = map(int, bbox)

    # Draw rectangle
    cv2.rectangle(image, (x1, y1), (x2, y2), color, thickness)

    # Draw label if provided
    if label:
        # Get text size
        (text_width, text_height), baseline = cv2.getTextSize(
            label, cv2.FONT_HERSHEY_SIMPLEX, 0.6, 2
        )

        # Draw label background
        cv2.rectangle(
            image,
            (x1, y1 - text_height - baseline - 5),
            (x1 + text_width, y1),
            color,
            -1
        )

        # Draw label text
        cv2.putText(
            image,
            label,
            (x1, y1 - 5),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            (255, 255, 255),
            2
        )

    return image


def draw_detections(image: np.ndarray,
                    boxes: np.ndarray,
                    classes: np.ndarray,
                    scores: np.ndarray,
                    labels: Optional[List[str]] = None,
                    color: Tuple[int, int, int] = (0, 255, 0),
                    thickness: int = 2) -> np.ndarray:
    """
    Draw multiple detection results on an image.

    Args:
        image: Input image
        boxes: Bounding boxes in [x1, y1, x2, y2] format, shape (N, 4)
        classes: Class indices, shape (N,)
        scores: Confidence scores, shape (N,)
        labels: Optional list of class labels
        color: Default box color
        thickness: Line thickness

    Returns:
        Image with drawn detections
    """
    result = image.copy()
    if len(boxes) == 0:
        return result
    for i in range(len(boxes)):
        box = boxes[i]
        class_id = int(classes[i])
        score = scores[i]

        # Prepare label text
        if labels and class_id < len(labels):
            label_text = f"{labels[class_id]} {score:.2f}"
        else:
            label_text = f"Class {class_id} {score:.2f}"

        result = draw_bbox(result, box, label_text, color, thickness)

    return result


def draw_tracking_results(image: np.ndarray,
                          tracks: List[Dict[str, Any]],
                          labels: Optional[List[str]] = None,
                          thickness: int = 2) -> np.ndarray:
    """
    Draw tracking results on an image.

    Args:
        image: Input image
        tracks: List of track dictionaries with keys:
                'tlbr': [x1, y1, x2, y2]
                'track_id': int
                'score': float
                'class_id': int (optional)
        labels: Optional list of class labels
        thickness: Line thickness

    Returns:
        Image with drawn tracks
    """
    result = image.copy()
    if not tracks:
        return result
    colors = generate_colors(100)

    for track in tracks:
        tlbr = track['tlbr']
        track_id = track['track_id']
        score = track['score']
        class_id = track.get('class_id', 0)

        x1, y1, x2, y2 = map(int, tlbr)

        # Get color based on track_id
        color = colors[track_id % len(colors)]

        # Draw bounding box
        cv2.rectangle(result, (x1, y1), (x2, y2), color, thickness)

        # Prepare label text
        if labels and class_id < len(labels):
            label_text = f"{labels[class_id]} ID:{track_id} {score:.2f}"
        else:
            label_text = f"ID:{track_id} {score:.2f}"

        # Draw label background
        (text_width, text_height), baseline = cv2.getTextSize(
            label_text, cv2.FONT_HERSHEY_SIMPLEX, 0.6, 2
        )
        cv2.rectangle(
            result,
            (x1, y1 - text_height - baseline - 5),
            (x1 + text_width, y1),
            color,
            -1
        )

        # Draw label text
        cv2.putText(
            result,
            label_text,
            (x1, y1 - 5),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            (255, 255, 255),
            2
        )

    return result


def draw_keypoints(
    image: np.ndarray,
    detections: List[Dict[str, Any]],
    box_color: Tuple[int, int, int] = (0, 255, 0),
    kp_color: Tuple[int, int, int] = (255, 0, 0),
    line_thickness: int = 2,
    kp_radius: int = 5,
    confidence_threshold: float = 0.2,
) -> np.ndarray:
    KP_CONNECTIONS = [
        [16, 14], [14, 12], [15, 13], [13, 11], [12, 11],
        [5, 7], [7, 9], [6, 8], [8, 10],
        [5, 6], [5, 11], [6, 12],
        [11, 13], [12, 14],
        [0, 1], [0, 2], [1, 3], [2, 4],
        [0, 5], [0, 6],
        [3, 5], [4, 6],
    ]

    for det in detections:
        # 绘制检测框
        box = det['box']
        x1, y1, x2, y2 = box
        cv2.rectangle(image, (x1, y1), (x2, y2), box_color, line_thickness)

        # 绘制关键点和连接线
        keypoints = det['keypoints']
        for i, (x, y, vis) in enumerate(keypoints):
            if vis < confidence_threshold:
                continue  # 跳过不可见关键点
            cv2.circle(image, (int(x), int(y)), kp_radius, kp_color, -1)

        # 绘制关键点连接线
        for (start_idx, end_idx) in KP_CONNECTIONS:
            if start_idx >= len(keypoints) or end_idx >= len(keypoints):
                continue  # 防止越界访问
            start_x, start_y, start_vis = keypoints[start_idx]
            end_x, end_y, end_vis = keypoints[end_idx]
            if start_vis < confidence_threshold or end_vis < confidence_threshold:
                continue
            cv2.line(
                image,
                (int(start_x), int(start_y)),
                (int(end_x), int(end_y)),
                kp_color,
                line_thickness,
            )

    return image


def draw_segmentation(
    image: np.ndarray,
    detections: List[Dict[str, Any]],
    masks: Optional[np.ndarray] = None,
    labels: List[str] = None,
) -> np.ndarray:
    """
    在图像上绘制检测框和分割掩码

    Args:
        image: 原始图像 (BGR格式)
        detections: 检测结果列表，每个元素包含 bbox, class, class_id, confidence
        masks: 分割掩码数组 [N, H, W]，可选
        labels: 类别标签列表

    Returns:
        绘制后的图像
    """
    result_image = image.copy()
    orig_h, orig_w = image.shape[:2]
    src_colors = [
        (4, 42, 255), (11, 219, 235), (243, 243, 243), (0, 223, 183),
        (17, 31, 104), (255, 111, 221), (255, 68, 79), (204, 237, 0),
        (0, 243, 68), (189, 0, 255), (0, 180, 255), (221, 0, 186),
        (0, 255, 255), (38, 192, 0), (1, 255, 179),
        (125, 36, 255), (123, 0, 104), (255, 27, 108), (252, 109, 47),
        (162, 255, 11),
    ]
    # 如果有掩码，先绘制掩码
    if masks is not None and len(masks) > 0:
        # 创建掩码叠加层
        overlay = result_image.copy()

        for i, mask in enumerate(masks):
            if i >= len(detections):
                break

            # 获取类别ID和颜色
            class_id = detections[i].get('class_id', 0)
            color_idx = class_id % len(src_colors)
            color = src_colors[color_idx]

            # 将掩码转换为3通道
            mask_3d = np.zeros((orig_h, orig_w, 3), dtype=np.uint8)
            mask_binary = (mask > 0.5).astype(np.uint8)
            mask_3d[:, :, 0] = mask_binary * color[0]
            mask_3d[:, :, 1] = mask_binary * color[1]
            mask_3d[:, :, 2] = mask_binary * color[2]

            # 叠加掩码
            overlay = cv2.addWeighted(overlay, 1.0, mask_3d, 0.5, 0)

        result_image = overlay

    # 绘制边界框和标签
    for i, det in enumerate(detections):
        bbox = det['bbox']  # [x1, y1, x2, y2]
        class_name = labels[det.get('class_id', 0)]
        confidence = det.get('confidence', 0.0)
        class_id = det.get('class_id', 0)

        # 确保坐标是整数且在图像范围内
        x1 = max(0, min(int(bbox[0]), orig_w - 1))
        y1 = max(0, min(int(bbox[1]), orig_h - 1))
        x2 = max(x1 + 1, min(int(bbox[2]), orig_w))
        y2 = max(y1 + 1, min(int(bbox[3]), orig_h))

        # 获取颜色
        color_idx = class_id % len(src_colors)
        color = src_colors[color_idx]

        # 绘制边界框
        cv2.rectangle(result_image, (x1, y1), (x2, y2), color, 2)

        # 准备标签文本
        label = f"{class_name} {confidence:.2f}"

        # 计算文本大小
        (text_width, text_height), baseline = cv2.getTextSize(
            label, cv2.FONT_HERSHEY_SIMPLEX, 0.6, 1
        )

        # 绘制文本背景
        text_y = max(y1, text_height + 10)
        cv2.rectangle(
            result_image,
            (x1, text_y - text_height - baseline - 5),
            (x1 + text_width, text_y + baseline),
            color,
            -1
        )

        # 绘制文本
        cv2.putText(
            result_image,
            label,
            (x1, text_y),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.6,
            (255, 255, 255) if sum(color) < 400 else (0, 0, 0),  # 根据背景色选择文本颜色
            1,
            cv2.LINE_AA
        )

    return result_image


def put_fps_text(image: np.ndarray,
                 fps: float,
                 position: Tuple[int, int] = (10, 30),
                 color: Tuple[int, int, int] = (0, 255, 0)) -> np.ndarray:
    """
    Draw FPS text on an image.

    Args:
        image: Input image
        fps: Frames per second
        position: Text position (x, y)
        color: Text color

    Returns:
        Image with FPS text
    """
    text = f"FPS: {fps:.1f}"
    cv2.putText(
        image,
        text,
        position,
        cv2.FONT_HERSHEY_SIMPLEX,
        1.0,
        color,
        2
    )
    return image

File: src/common/cpp/drawing.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "drawing.h"

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

namespace vision_common {

void draw_detections(
    cv::Mat& image,
    const std::vector<vision_common::DetectionResult>& detections,
    const std::vector<std::string>& labels,
    const cv::Scalar& box_color,
    int line_thickness) {
    if (detections.empty()) return;
    for (const auto& det : detections) {
        int ix1 = static_cast<int>(det.bbox.x1);
        int iy1 = static_cast<int>(det.bbox.y1);
        int ix2 = static_cast<int>(det.bbox.x2);
        int iy2 = static_cast<int>(det.bbox.y2);

        // Draw bounding box
        cv::rectangle(image, cv::Point(ix1, iy1), cv::Point(ix2, iy2), box_color, line_thickness);

        // Draw label and score (skip if class_id is -1, e.g., for face detection without labels)
        if (det.label >= 0) {
            std::string labelText;
            if (!labels.empty() && det.label < static_cast<int>(labels.size())) {
                labelText = labels[det.label] + ": " + std::to_string(det.score).substr(0, 4);
            } else {
                labelText = "Class " + std::to_string(det.label) + ": " + std::to_string(det.score).substr(0, 4);
            }

            cv::putText(image, labelText, cv::Point(ix1, iy1 - 10),
                        cv::FONT_HERSHEY_SIMPLEX, 0.9, box_color, line_thickness);
        }
    }
}

void draw_keypoints(
    cv::Mat& image,
    const std::vector<vision_common::PoseResult>& results,
    float point_confidence_threshold,
    const cv::Scalar& box_color,
    const cv::Scalar& kp_color,
    int line_thickness,
    int kp_radius) {
    if (results.empty()) return;
    // Keypoint connections for pose estimation (COCO format: 17 keypoints)
    // These connections define the skeleton structure
    std::vector<std::pair<int, int>> kp_connections = {
        {16, 14}, {14, 12}, {15, 13}, {13, 11}, {12, 11},  // Head to shoulders
        {5, 7}, {7, 9}, {6, 8}, {8, 10},                   // Arms
        {5, 6}, {5, 11}, {6, 12},                          // Shoulders to torso
        {11, 13}, {12, 14},                                 // Torso to hips
        {0, 1}, {0, 2}, {1, 3}, {2, 4},                    // Face
        {0, 5}, {0, 6},                                     // Face to shoulders
        {3, 5}, {4, 6}  // Face to shoulders (alternate)
    };

    for (const auto& result : results) {
        int x1 = static_cast<int>(result.bbox.x1);
        int y1 = static_cast<int>(result.bbox.y1);
        int x2 = static_cast<int>(result.bbox.x2);
        int y2 = static_cast<int>(result.bbox.y2);

        // Draw bounding box
        cv::rectangle(image, cv::Point(x1, y1), cv::Point(x2, y2), box_color, line_thickness);

        // Draw label and score
        std::string labelText = "Person: " + std::to_string(result.score).substr(0, 4);
        cv::putText(image, labelText, cv::Point(x1, y1 - 10),
                    cv::FONT_HERSHEY_SIMPLEX, 0.9, box_color, line_thickness);

        // Draw keypoints
        for (size_t i = 0; i < result.keypoints.size(); ++i) {
            const auto& kp = result.keypoints[i];
            if (kp.visibility < point_confidence_threshold) {
                continue;  // Skip invisible keypoints
            }
            // Draw keypoint circle
            cv::circle(image, cv::Point(static_cast<int>(kp.x), static_cast<int>(kp.y)),
                        kp_radius, kp_color, -1);
        }

        // Draw keypoint connections (skeleton)
        for (const auto& connection : kp_connections) {
            int start_idx = connection.first;
            int end_idx = connection.second;

            // Check bounds
            if (start_idx >= static_cast<int>(result.keypoints.size()) ||
                end_idx >= static_cast<int>(result.keypoints.size())) {
                continue;
            }

            const auto& start_kp = result.keypoints[start_idx];
            const auto& end_kp = result.keypoints[end_idx];

            // Only draw line if both keypoints are visible
            if (start_kp.visibility < point_confidence_threshold ||
                end_kp.visibility < point_confidence_threshold) {
                continue;
            }

            cv::line(image,
                    cv::Point(static_cast<int>(start_kp.x), static_cast<int>(start_kp.y)),
                    cv::Point(static_cast<int>(end_kp.x), static_cast<int>(end_kp.y)),
                    kp_color, line_thickness);
        }
    }
}

void draw_segmentation(
    cv::Mat& image,
    const std::vector<vision_common::SegmentationResult>& results,
    const std::vector<std::string>& labels,
    float alpha,
    const cv::Scalar& box_color,
    int line_thickness) {
    if (results.empty()) return;
    // Fixed palette for stable and visually pleasing mask colors (BGR).
    static const std::vector<cv::Scalar> kPalette = {
        cv::Scalar(56, 56, 255),   cv::Scalar(151, 157, 255), cv::Scalar(31, 112, 255),
        cv::Scalar(29, 178, 255),  cv::Scalar(49, 210, 207),  cv::Scalar(10, 249, 72),
        cv::Scalar(23, 204, 146),  cv::Scalar(134, 219, 61),  cv::Scalar(52, 147, 26),
        cv::Scalar(187, 212, 0),   cv::Scalar(212, 188, 0),   cv::Scalar(255, 157, 151),
        cv::Scalar(255, 56, 132),  cv::Scalar(255, 102, 187), cv::Scalar(255, 149, 200),
        cv::Scalar(44, 153, 168),  cv::Scalar(0, 194, 255),   cv::Scalar(52, 69, 147),
        cv::Scalar(100, 115, 255), cv::Scalar(152, 251, 152)
    };

    // Create overlay image for masks only
    cv::Mat overlay = image.clone();

    for (size_t i = 0; i < results.size(); ++i) {
        const auto& result = results[i];
        const int color_index = (result.label >= 0)
            ? (result.label % static_cast<int>(kPalette.size()))
            : (static_cast<int>(i) % static_cast<int>(kPalette.size()));
        cv::Scalar color = kPalette[static_cast<size_t>(color_index)];

        // Draw mask if available
        if (result.mask != nullptr && !result.mask->empty()) {
            cv::Mat mask = *result.mask;

            // Ensure mask is CV_8U type (required by setTo)
            if (mask.type() != CV_8U) {
                cv::Mat mask_uint8;
                if (mask.type() == CV_32F) {
                    // Convert from float [0, 1] to uint8 [0, 255]
                    mask.convertTo(mask_uint8, CV_8U, 255.0);
                } else {
                    mask.convertTo(mask_uint8, CV_8U);
                }
                mask = mask_uint8;
            }

            // Resize mask if needed (must match image size)
            if (mask.size() != image.size()) {
                cv::Mat mask_resized;
                cv::resize(mask, mask_resized, image.size(), 0, 0, cv::INTER_NEAREST);
                mask = mask_resized;
            }

            // Ensure mask is single channel (required by setTo)
            if (mask.channels() != 1) {
                cv::Mat mask_single;
                if (mask.channels() == 3) {
                    cv::cvtColor(mask, mask_single, cv::COLOR_BGR2GRAY);
                } else {
                    // Extract first channel
                    std::vector<cv::Mat> channels;
                    cv::split(mask, channels);
                    mask_single = channels[0];
                }
                mask = mask_single;
            }

            // Apply mask to overlay (mask must be CV_8U, single channel, same size as overlay)
            overlay.setTo(color, mask);
        }
    }

    // Blend overlay with original image, then draw boxes/labels on top at full opacity
    cv::addWeighted(image, 1.0 - alpha, overlay, alpha, 0, image);

    for (size_t i = 0; i < results.size(); ++i) {
        const auto& result = results[i];
        // Negative bbox (e.g. x1 < 0) skips box/label overlay for mask-only semantic layers.
        if (result.bbox.x1 < 0.0f) {
            continue;
        }

        int x1 = static_cast<int>(result.bbox.x1);
        int y1 = static_cast<int>(result.bbox.y1);
        int x2 = static_cast<int>(result.bbox.x2);
        int y2 = static_cast<int>(result.bbox.y2);

        cv::rectangle(image, cv::Point(x1, y1), cv::Point(x2, y2), box_color, line_thickness);

        if (result.label >= 0) {
            std::string labelText;
            if (!labels.empty() && result.label < static_cast<int>(labels.size())) {
                labelText = labels[result.label] + ": " + std::to_string(result.score).substr(0, 4);
            } else {
                labelText = "Class " + std::to_string(result.label) + ": " + std::to_string(result.score).substr(0, 4);
            }
            int label_y = std::max(2, y1 - 10);
            cv::putText(image, labelText, cv::Point(x1, label_y),
                        cv::FONT_HERSHEY_SIMPLEX, 0.9, box_color, line_thickness);
        }
    }
}

/**
 * @brief Get color for track ID
 */
cv::Scalar get_track_color(int track_id) {
    // Use a simple hash to generate consistent colors for each track ID
    static const int colors[][3] = {
        {255, 0, 0}, {0, 255, 0}, {0, 0, 255}, {255, 255, 0},
        {255, 0, 255}, {0, 255, 255}, {128, 0, 0}, {0, 128, 0},
        {0, 0, 128}, {128, 128, 0}, {128, 0, 128}, {0, 128, 128}
    };
    constexpr int num_colors = 12;
    int color_idx = (track_id % num_colors + num_colors) % num_colors;
    return cv::Scalar(colors[color_idx][0], colors[color_idx][1], colors[color_idx][2]);
}

void draw_tracking_results(
    cv::Mat& image,
    const std::vector<TrackingResult>& results,
    const std::vector<std::string>& labels,
    int line_thickness) {
    if (results.empty()) return;
    for (const auto& result : results) {
        // Get color for this track
        cv::Scalar color = get_track_color(result.track_id);

        // Draw bounding box
        cv::rectangle(image,
                        cv::Point(static_cast<int>(result.bbox.x1), static_cast<int>(result.bbox.y1)),
                        cv::Point(static_cast<int>(result.bbox.x2), static_cast<int>(result.bbox.y2)),
                        color, line_thickness);

        // Prepare label text: class ID score (one line)
        std::string label_text;
        if (result.label >= 0 && result.label < static_cast<int>(labels.size())) {
            label_text = labels[result.label] + " ID:" + std::to_string(result.track_id);
        } else {
            label_text = "ID:" + std::to_string(result.track_id);
        }
        label_text += " " + std::to_string(static_cast<int>(result.score * 100)) + "%";

        // Draw label background (clamp to image to avoid negative coordinates)
        int baseline = 0;
        cv::Size text_size = cv::getTextSize(label_text, cv::FONT_HERSHEY_SIMPLEX, 0.5, 1, &baseline);
        int label_y_top = std::max(0, static_cast<int>(result.bbox.y1) - text_size.height - 5);
        int label_y_bottom = static_cast<int>(result.bbox.y1);
        cv::rectangle(image,
                        cv::Point(static_cast<int>(result.bbox.x1), label_y_top),
                        cv::Point(static_cast<int>(result.bbox.x1) + text_size.width, label_y_bottom),
                        color, -1);

        // Draw label text
        cv::putText(image, label_text,
                        cv::Point(static_cast<int>(result.bbox.x1), std::max(2, static_cast<int>(result.bbox.y1) - 5)),
                        cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(255, 255, 255), 1);
    }
}

// Generic draw function that handles ModelResult variant
void draw_results(
    cv::Mat& image,
    const std::vector<ModelResult>& results,
    const std::vector<std::string>& labels,
    int line_thickness) {
    if (results.empty()) return;

    // Separate results by type
    std::vector<DetectionResult> detections;
    std::vector<PoseResult> poses;
    std::vector<SegmentationResult> segmentations;
    std::vector<TrackingResult> trackings;
    std::vector<TextResult> texts;
    std::vector<DisparityResult> disparities;
    std::vector<LocalFeaturesResult> local_features;

    for (const auto& result : results) {
        std::visit([&](const auto& r) {
            using T = std::decay_t<decltype(r)>;
            if constexpr (std::is_same_v<T, DetectionResult>) {
                detections.push_back(r);
            } else if constexpr (std::is_same_v<T, PoseResult>) {
                poses.push_back(r);
            } else if constexpr (std::is_same_v<T, SegmentationResult>) {
                segmentations.push_back(r);
            } else if constexpr (std::is_same_v<T, TrackingResult>) {
                trackings.push_back(r);
            } else if constexpr (std::is_same_v<T, TextResult>) {
                texts.push_back(r);
            } else if constexpr (std::is_same_v<T, DisparityResult>) {
                disparities.push_back(r);
            } else if constexpr (
                std::is_same_v<T, LocalFeaturesResult>) {
                local_features.push_back(r);
            }
        }, result);
    }

    if (!disparities.empty()) {
        const auto& disparity = disparities.front();
        if (disparity.map == nullptr || disparity.map->empty() ||
            disparity.map->type() != CV_32FC1) {
            throw std::runtime_error(
                "disparity draw expects a non-empty CV_32FC1 map");
        }
        const cv::Mat& map = *disparity.map;
        float min_value = std::numeric_limits<float>::infinity();
        float max_value = -std::numeric_limits<float>::infinity();
        for (int y = 0; y < map.rows; ++y) {
            const float* row = map.ptr<float>(y);
            for (int x = 0; x < map.cols; ++x) {
                if (!std::isfinite(row[x])) {
                    continue;
                }
                min_value = std::min(min_value, row[x]);
                max_value = std::max(max_value, row[x]);
            }
        }
        if (!std::isfinite(min_value) || !std::isfinite(max_value)) {
            throw std::runtime_error(
                "disparity draw requires at least one finite value");
        }
        cv::Mat normalized(map.size(), CV_8UC1, cv::Scalar(0));
        const float scale = max_value > min_value
            ? 255.0f / (max_value - min_value)
            : 0.0f;
        for (int y = 0; y < map.rows; ++y) {
            const float* source = map.ptr<float>(y);
            uint8_t* destination = normalized.ptr<uint8_t>(y);
            for (int x = 0; x < map.cols; ++x) {
                if (std::isfinite(source[x])) {
                    destination[x] = cv::saturate_cast<uint8_t>(
                        (source[x] - min_value) * scale);
                }
            }
        }
        cv::applyColorMap(normalized, image, cv::COLORMAP_JET);
    }
    for (const auto& features : local_features) {
        for (const auto& point : features.keypoints) {
            if (!std::isfinite(point.x) || !std::isfinite(point.y) ||
                !std::isfinite(point.visibility) ||
                point.visibility <= 0.0f ||
                point.x < 0.0f || point.y < 0.0f ||
                point.x >= static_cast<float>(image.cols) ||
                point.y >= static_cast<float>(image.rows)) {
                continue;
            }
            cv::circle(
                image,
                cv::Point(
                    static_cast<int>(std::round(point.x)),
                    static_cast<int>(std::round(point.y))),
                2,
                cv::Scalar(0, 255, 0),
                -1);
        }
    }

    // Draw each type with appropriate function
    if (!detections.empty()) {
        draw_detections(image, detections, labels, cv::Scalar(0, 255, 0), line_thickness);
    }
    if (!poses.empty()) {
        draw_keypoints(image, poses, 0.2f, cv::Scalar(0, 255, 0), cv::Scalar(255, 0, 0), line_thickness, 5);
    }
    if (!segmentations.empty()) {
        draw_segmentation(image, segmentations, labels, 0.5f, cv::Scalar(0, 255, 0), line_thickness);
    }
    if (!trackings.empty()) {
        draw_tracking_results(image, trackings, labels, line_thickness);
    }
    if (!texts.empty()) {
        draw_text(image, texts, cv::Scalar(0, 255, 0), line_thickness);
    }
}

void draw_text(
    cv::Mat& image,
    const std::vector<vision_common::TextResult>& results,
    const cv::Scalar& color,
    int line_thickness) {
    for (const auto& r : results) {
        if (r.polygon.size() < 2) {
            continue;
        }
        // Draw the polygon outline (usually a quadrilateral).
        std::vector<cv::Point> pts;
        pts.reserve(r.polygon.size());
        for (const auto& kp : r.polygon) {
            pts.emplace_back(static_cast<int>(kp.x), static_cast<int>(kp.y));
        }
        const cv::Point* p = pts.data();
        int n = static_cast<int>(pts.size());
        cv::polylines(image, &p, &n, 1, true, color, line_thickness);

        // Put the recognized text above the first (top-left) corner.
        // NOTE: cv::putText renders ASCII only; non-ASCII (e.g. Chinese) shows
        // as '?'. The recognized string itself is intact in the result.
        const std::string label = r.text + " " + std::to_string(r.score).substr(0, 4);
        const int tx = pts[0].x;
        const int ty = std::max(pts[0].y - 5, 12);
        cv::putText(image, label, cv::Point(tx, ty), cv::FONT_HERSHEY_SIMPLEX, 0.6,
                    color, line_thickness);
    }
}

}  // namespace vision_common
